tractolearn/generative/filter_generated_points.py

```diff
@@ -0,0 +1,657 @@
+# -*- coding: utf-8 -*-
+import logging
+from enum import Enum
+
+import numpy as np
+import torch
+from dipy.io.stateful_tractogram import StatefulTractogram
+from nibabel.streamlines import ArraySequence
+from scilpy.segment.streamlines import streamlines_in_mask
+from scipy.ndimage import map_coordinates
+
+from tractolearn.tractomath.utils import unit_vector, compute_angle
+from tractolearn.transformation.peaks_utils import (
+    get_peak_count,
+    reshape_peaks_for_computing,
+)
+from tractolearn.transformation.streamline_transformation import resample_streamlines
+from tractolearn.transformation.volume_utils import interpolate_volume_at_coordinates
+
+torch.set_flush_denormal(True)
+logger = logging.getLogger("root")
+
+
+class StreamlineFeatures(Enum):
+    FA = "fa"
+    LENGTH = "length"
+    LOCAL_ORIENTATION_ANGLE = "local_orientation_angle"
+    MEAN_CURVATURE = "mean_curvature"
+    REGION = "region"
+    WINDING = "winding"
+    WM_OCCUPANCY = "wm_occupancy"
+    GM_OCCUPANCY = "gm_occupancy"
+    WM_MEMBERSHIP = "wm_membership"
+    WM_VOXEL_OCCUPANCY_RATIO = "wm_voxel_occupancy_ratio"
+    CORTICAL_GAP = "cortical_gap"
+    SURFACE_INTERSECTION = "surface_intersection"
+    SURFACE_INCIDENCE_ANGLE = "surface_incidence_angle"
+    SURFACE_PROJECTED_ANGLE = "surface_projected_angle"
+
+
+def filter_grid_roi(sft, mask, filter_type, is_exclude, soft_percentage: float = None):
+    """
+    Parameters
+    ----------
+    sft : StatefulTractogram
+        StatefulTractogram containing the streamlines to segment.
+    target_mask : numpy.ndarray
+        Binary mask in which the streamlines should pass.
+    filter_type: str
+        One of the 3 following choices, 'any', 'all', 'either_end', 'both_ends'.
+    is_exclude: bool
+        Value to indicate if the ROI is an AND (false) or a NOT (true).
+    Returns
+    -------
+    ids : tuple
+        Filtered sft.
+        Ids of the streamlines passing through the mask.
+    """
+    line_based_indices = []
+    if filter_type in ["any", "all"]:
+        line_based_indices = streamlines_in_mask(sft, mask, all_in=filter_type == "all")
+    else:
+        sft.to_vox()
+        sft.to_corner()
```

Comment:
Should be put into a function to always do the same thing, in the same manner, and to avoid forgetting one of the two. Same applies to undoing it.

Reply:
Done